alf/algorithms/agent.py

```diff
@@ -142,6 +154,14 @@ def predict(self, time_step: ActionTimeStep, state: AgentState,
         observation = self._encode(time_step)
 
         new_state = AgentState()
+        if self._goal_generator is not None:
+            goal_step = self._goal_generator.rollout(
+                observation=observation,
+                state=state.goal_generator,
+                step_type=time_step.step_type)
+            new_state = new_state._replace(goal_generator=goal_step.state)
+
+            observation = [observation, goal_step.state]
```

Comment:
It's better not to use state as goal. It's quite possible that state can contain other information for a different future goal generator. Perhaps goal_step.action (if changing goal generator to be an RLAlgorithm). And we can save the goal to to info so that it can be used at train_step

Reply:
Code has been updated to avoid using goal_step.states as the goal vector directly.